Update: object-curly-spacing spaceInEmptyObject option (fixes #7927)

pmcelhaney opened the pull request:
<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

https://github.com/eslint/eslint/issues/7927

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (http://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

**What changes did you make? (Give an overview)**

Added a new option, "spaceInEmptyObject". It can be "never" or "always". The default is to ignore empty objects, as the rule does now. 


**Is there anything you'd like reviewers to focus on?**

No.

platinumazure commented:
@eslint/eslint-team Issue #7927 was originally accepted as a bug, where the resolution seemed to be that `object-curly-spacing: ["error", "always"]` should enforce `{ }` rather than `{}`. This PR is instead an enhancement to add an optional option that can enforce one way or the other, and the absence of the option will allow both styles for empty objects. I think we need to evaluate this enhancement request separately from the original bug, but bearing in mind that it should be a good option for fixing that bug.

I'll champion this. We need three :+1:s from other team members. Let's do this.

platinumazure commented on lib/rules/object-curly-spacing.js:
```diff
@@ -32,6 +32,9 @@ module.exports = {
                         type: "boolean"
+                    },
+                    spaceInEmptyObject: {
+                        type: "string"
```
I think this should be an `enum` so we can enforce always/never specifically. (See line 25 for another example)

platinumazure commented on tests/lib/rules/object-curly-spacing.js:
```diff
@@ -58,6 +58,9 @@ ruleTester.run("object-curly-spacing", rule, {
         // always - empty object
         { code: "var foo = {};", options: ["always"] },
+        { code: "var foo = {};", options: ["always", { spaceInEmptyObject: "never" }] },
+        { code: "var foo = { };", options: ["always", { spaceInEmptyObject: "always" }] },
```
Could you add some tests for line breaks? (I assume those should always pass and should be handled by the `object-curly-newline` rule.)

```js
var foo = {
};
// Should pass(?) for all combinations of first option + spaceInEmptyObject option
```

not-an-aardvark commented on docs/rules/object-curly-spacing.md:
```diff
@@ -37,6 +37,8 @@ Object option:
 * `"objectsInObjects": true` requires spacing inside of braces of objects beginning and/or ending with an object element (applies when the first option is set to `never`)
 * `"objectsInObjects": false` disallows spacing inside of braces of objects beginning and/or ending with an object element (applies when the first option is set to `always`)
+* `"spaceInEmptyObject": "always"`: requires a space inside an empty object literal (`{ }` is valid, `{}` is invalid)
+* `"spaceInEmptyObject": "never"`: disallows a space inside an empty object literal (`{}` is valid, `{ }` is invalid)
```
For consistency with the other options, can we change the name and schema to `emptyObjects: true` or `emptyObjects: false`?

pmcelhaney replied:
I'm going for consistency with the main option.

not-an-aardvark replied:
Thanks for explaining. It looks like there is already some inconsistency here (between `always`/`never` for the main option, and `true`/`false` for the object options). Having said that, I still think it would be good to maintain consistency *among* the object options, so that we don't end up with one object option that arbitrarily has different semantics from the other options. To that end, I still think it would be better to use `true`/`false` here, and to give it a plural name without the word `space` (such as `emptyObjects`).

platinumazure replied:
I can see the argument for consistency, but personally I think always/never is more clear. True/false could be read as either, enforce space/enforce no space (i.e., equivalent to always/never), or it could be read as enforce in this case/don't enforce in this case (i.e., false = ignore, and true would mean always or never specifically).

I'm willing to bow towards consistency, but I think we should seriously reconsider the use of boolean options when the false could be read to mean either enforcing a lack of something, or not enforcing something.

not-an-aardvark replied:
I would also be fine with changing all of the options to accept `always`/`never` (and deprecating the boolean versions). I just don't want to enforce one convention for one option, and a different convention for another option.

pmcelhaney replied:
Before making this change, I want to make sure we're clear.

These values will remain for backward compatibility. At least until the next major release, maybe longer.

> * `"objectsInObjects": true` requires spacing inside of braces of objects beginning and/or ending with an object element (applies when the first option is set to `never`)
 > * `"objectsInObjects": false` disallows spacing inside of braces of objects beginning and/or ending with an object element (applies when the first option is set to `always`)

These new values will be added. Note the big difference is "objectsInObjects" et. al. are now decoupled from the first option. 

 > * `"objectsInObjects": "always"` requires spacing inside of braces of objects beginning and/or ending with an object element 
 > * `"objectsInObjects": "never"` disallows spacing inside of braces of objects beginning and/or ending with an object element 

Should I bring this up in the issue thread? I'd like to have consensus before moving forward.

not-an-aardvark replied:
If we're going to allow `always`/`never` as aliases for `true`/`false` for all the options, I think we might want to make that a separate issue and PR so that we don't block adding this option. My preference would be to use `true`/`false` with `emptyObjects` for now so that we can remain consistent, and then consider allowing `always`/`never`  for all the options separately.

not-an-aardvark commented on tests/lib/rules/object-curly-spacing.js:
```diff
@@ -120,6 +123,8 @@ ruleTester.run("object-curly-spacing", rule, {
         // never - empty object
         { code: "var foo = {};", options: ["never"] },
+        { code: "var foo = {};", options: ["never", { spaceInEmptyObject: "never" }] },
+        { code: "var foo = { };", options: ["never", { spaceInEmptyObject: "always" }] },
```
Can you add a test where an empty object contains a comment?

```js
var x = { /* no properties */ }
```

pmcelhaney commented:
@platinumazure @not-an-aardvark Thanks for the quick feedback! I've added tests for comments, new lines and a mix of various white space. Also changed the option to an emum.

not-an-aardvark commented on tests/lib/rules/object-curly-spacing.js:
```diff
@@ -120,6 +125,12 @@ ruleTester.run("object-curly-spacing", rule, {
         { code: "var foo = {};", options: ["never"] },
+        { code: "var foo = {};", options: ["never", { spaceInEmptyObject: "never" }] },
+        { code: "var foo = { };", options: ["never", { spaceInEmptyObject: "always" }] },
+        { code: "var foo = {/* this is a comment */};", options: ["never", { spaceInEmptyObject: "always" }] },
```
I think this case will currently report an error with `spaceInEmptyObject: "never"`, but I don't think that's desirable -- the rule shouldn't forbid people from putting comments in objects. Instead, I would expect the rule to enforce or forbid spaces *around* the comment in this case, depending on either the main option or the `spaceInEmptyObject` option.

pmcelhaney replied:
Are you saying it *should* report an error? I don't think the ~~"spaceInEmptyObject~~ "emptyObjects" option should have any bearing on an object that contains anything other than whitespace. 

Instead, we need yet another option for "commentsInObjects".

not-an-aardvark replied:
I'm saying that the following code *should not* report an error, but it does report an error with the current implementation:

```js
/* eslint object-curly-spacing: [error, never, { spaceInEmptyObject: never }] */

var foo = {/* this is a comment */};
```

The issue with the current behavior is that `spaceInEmptyObject: "never"` is unsatisfiable for objects with a comment and no properties, because the rule still measures the offset between the `{` and the `}` rather than measuring the offset between the `{` and the start of the comment.

I don't feel strongly about how cases like the following should be treated:

```js
/* eslint object-curly-spacing: [error, always, { spaceInEmptyObject: never }] */

var foo = {/* this is a comment */};
```

I think I would want an error to be reported since there is no space between the `{` and the `/*`, and the object isn't "empty", but I'm open to being convinced otherwise.

pmcelhaney replied:
Okay, yes, that should not report an error.

platinumazure commented:
When considering empty objects, should we define those by lack of tokens/comments between the curly braces, instead of lack of properties? I feel like that makes more sense conceptually. What do you guys think?

not-an-aardvark commented:
Sounds good to me. However, that could cause this change to start reporting more errors if someone currently has code like this:

```js
/* eslint object-curly-spacing: [error, always] */

({/* comment */})
```

platinumazure commented:
@eslint/eslint-team One more :+1: and we could accept this enhancement!

btmills commented:
Just added my :+1: - this is now accepted!

not-an-aardvark commented:
@pmcelhaney Friendly ping; do you think you will be able to fix the issue described in https://github.com/eslint/eslint/pull/9244#discussion_r137401272?

pmcelhaney commented:
Yes. Sorry, didn’t realize / forgot there was work left to be done. Are you referring to the comment to which I replied, “Okay, yes, that should not report an error”?

not-an-aardvark commented:
Yes, that's what I was referring to.

pmcelhaney commented on tests/lib/rules/object-curly-spacing.js:
```diff
@@ -811,6 +811,17 @@ ruleTester.run("object-curly-spacing", rule, {
         },
+        {
+            code: "var foo = { /* comment */ };",
+            output: "var foo = {};",
```
Oh no, that's no good. We don't want to delete the comment! I'll work on either removing the spaces and leaving the comment or skipping the fix in this case.

pmcelhaney commented on tests/lib/rules/object-curly-spacing.js:
```diff
@@ -58,6 +58,11 @@ ruleTester.run("object-curly-spacing", rule, {
+        { code: "var foo = {};", options: ["always", { spaceInEmptyObject: "never" }] },
+        { code: "var foo = { };", options: ["always", { spaceInEmptyObject: "always" }] },
+        { code: "var foo = { /* this is a comment */ };", options: ["always", { spaceInEmptyObject: "always" }] },
+        { code: "var foo = {\n};", options: ["always", { spaceInEmptyObject: "always" }] },
```
Fixed so that a comment is not mistaken for a space.

platinumazure commented:
Should the new option also handle space before the closing bracket? It looks like it might not right now based on my (admittedly very quick) skimming of the tests, but let me know if I'm mistaken. And if we aren't handling that space before the closing bracket-- should we?

pmcelhaney commented:
In an empty object, the space preceding the closing bracket is the same as the space following the opening bracket. 

*Unless* one or more comment between the brackets break up the space(s). 

Having given it some more thought, it’s probably best not to trigger the rule at all if a comment is found within an empty object. That may come into conflict with a rule about spacing around comments.

platinumazure commented:
> Having given it some more thought, it’s probably best not to trigger the rule at all if a comment is found within an empty object. That may come into conflict with a rule about spacing around comments.

This sounds like a sensible direction from this PR, and if we decide it makes sense to enforce in this rule, we can always change it later. :+1:

platinumazure commented:
Is this ready for review from not-an-aardvark again, or do you still need to address some of his comments? (Sorry, it's a bit hard to tell in this interface with all of the comments he raised.)

pmcelhaney commented:
Okay, I need to make some changes so that an empty object that contains a comment never triggers this rule. Will comment here when I'm done.

pmcelhaney commented:
I'm going to close this PR. I think we should set it aside for now. The difference between the semantics of `true`/`false` and `"always"`/`"never"` is too confusing. 

I need to think about it some more, but I'm leaning toward approaching the change in three stages. 

1. Add tests for `{ }` to make sure the behavior for existing configs is not impacted.
2. Add `"always"`/`"never"` options for `objectsInObjects` and `arraysInObjects` and deprecate `true`/`false`. 
3. Apply this change, adding `spaceInEmptyObjects` with support for `"always"/"never"` only.
